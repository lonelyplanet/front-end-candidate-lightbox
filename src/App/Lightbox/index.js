import React, { Component } from 'react'
import 'whatwg-fetch'
import './index.scss'

const API_KEY = 'MTgwNzAw'

class Lightbox extends Component {
  constructor (props) {
    super(props)

    this.state = { url: '', loading: false, error: '' }

    this.handleKeydown = this.handleKeydown.bind(this)
    this.handleClick = this.handleClick.bind(this)
    this.handleCloseBtn = this.handleCloseBtn.bind(this)
  }

  componentWillMount () {
    this.setState({ loading: true })

    fetch(`http://thecatapi.com/api/images/get?api_key=${API_KEY}`)
      .then((response) => {
        this.setState({ url: response.url, loading: false })
      })
      .catch(() => {
        this.setState({ loading: false, error: 'Unable to get a cat. Try again' })
      })

    document.addEventListener('keydown', this.handleKeydown);
    document.addEventListener('click', this.handleClick);
  }

  componentWillUnmount () {
    document.removeEventListener('keydown', this.handleKeydown);
    document.removeEventListener('click', this.handleClick);
  }

  handleClick (event) {
    if (event.target.className == 'lightbox') {
      this.props.onClose()
    }
  }

  handleCloseBtn (event) {
    event.preventDefault()
    this.props.onClose()
  }

  handleKeydown (event) {
    if (event.keyCode === 27) {
      this.props.onClose()
    }
  }

  render () {
    const loadingMessage = this.state.loading
      ? (<h3>Loading Catâ€¦</h3>)
      : null

    const errorMessage = this.state.error
      ? (<h4>{this.state.error}</h4>)
      : null

    return (
      <div className='lightbox'>
        <div className='lightbox-content'>
          <img src={this.state.url} />
          {loadingMessage}
          {errorMessage}
          <a href className="button" onClick={this.handleCloseBtn}>Close</a>
        </div>
      </div>
    )
  }
}

export default Lightbox
