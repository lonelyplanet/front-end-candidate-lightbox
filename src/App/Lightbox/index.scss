.lightbox {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background: rgba(0, 0, 0, 0.2);
  display: flex;
  justify-content: center;
  align-items: center;
}

.lightbox-content {
  background: #ffffff;
  width: 30vw;
  padding: 1rem;

  img {
    max-width: 100%;
    display: block;
    margin: 0 auto 1rem;
  }
}