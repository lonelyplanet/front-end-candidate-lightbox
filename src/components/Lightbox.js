import React, { Component, PropTypes } from 'react';
import cssModules from 'react-css-modules';
import Modal from 'react-modal';

const styles = require('../styles/lightbox.css');

const customStyles = {
  content : {
    top                   : '50%',
    left                  : '50%',
    right                 : 'auto',
    bottom                : 'auto',
    marginRight           : '-50%',
    transform             : 'translate(-50%, -50%)'
  }
};

class Lightbox extends Component {
  constructor(props) {
    super(props);

  }

  render() {
    const { isOpen } = this.props;
    const noop = () => {};

    return (
      <div styleName="lightbox">
        <Modal
          isOpen={isOpen}
          onRequestClose={noop}
          onAfterOpen={() => this.outerEl && this.outerEl.focus()}
          style={customStyles}
        >
          <div>I am a modal!!!</div>
        </Modal>
      </div>
    );
  }
}

Lightbox.defaultProps = {

};

Lightbox.propTypes = {

};

export default cssModules(Lightbox, styles);
