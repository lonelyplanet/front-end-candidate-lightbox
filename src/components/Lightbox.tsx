import * as React from 'react'
import * as cx from 'classnames'

export interface LightboxProps {
  onClose: () => any
  imageSrc: string
  title?: string
  description?: string
}
interface LightboxState {
  isLoading: boolean
  transition: 'enter' | 'leave' | null
}
class Lightbox extends React.Component<LightboxProps> {
  /** Ref to the lightbox DOM node */
  _lightbox: HTMLDivElement | null

  state: LightboxState = {
    // NOTE(zuko): brief point of discussion: We could set this to `false` as
    // the sane initial value and then immediately change it to `true` in
    // `componentWillMount` (or the constructor), but re: `componentWillMount`
    // I've heard a lot of discussion about that particularl lifecycle event
    // potentially being fired more than once per component instance once React
    // enables asynchronous rendering with Fiber. In that case, we would not
    // want one-time initialization logic run multiple times, and I'm not yet
    // convinced that it's better to move it to the constructor.
    //
    // For simplicity's sake I could just do what has been described in the
    // comment above, but lately I've been trying to future-proof my React
    // code and felt it warranted noting. I could very well wrong in this.
    isLoading: true,
    transition: 'enter',
  }

  componentDidMount() {
    document.body.classList.add('lightbox-open')
    document.addEventListener('click', this._checkForOutsideClick)
    document.addEventListener('keydown', this._checkKeydownEvent)
  }

  componentWillUnmount() {
    document.body.classList.remove('lightbox-open')
    document.removeEventListener('click', this._checkForOutsideClick)
    document.removeEventListener('keydown', this._checkKeydownEvent)
  }

  /**
   * Handles all click events that bubble up to the document root. If the
   * click event did not originate from within the lightbox, then we treat
   * that as an intent to close the lightbox.
   */
  private _checkForOutsideClick = (e: any) => {
    if (this._lightbox && !this._lightbox.contains(e.target)) {
      this._onClose()
    }
  }

  /**
   * Handles all keydown events that bubble up to the document root. If the
   * Escape key is pressed, close the lightbox.
   */
  private _checkKeydownEvent = (e: any) => {
    switch (e.key) {
      case 'Escape':
        this._onClose()
        break
      default:
      // noop
    }
  }

  private _onClose = () => {
    if (this.state.transition === 'leave') return

    this.setState({ transition: 'leave' }, () => {
      // Wait for the hide animation to complete. This is unfortunate coupling
      // between the component and its CSS, since we have to keep the animation
      // durations consistent. This could likely be improved by either setting
      // the animation duration directly from the JS or by incorporating more
      // modern CSS-in-JS solutions. For now, I'm just keeping it simple.
      setTimeout(() => {
        this.props.onClose()
      }, 250)
    })
    this._onClose()
  }

  private _onLightboxRef = (node: HTMLDivElement) => {
    this._lightbox = node
  }

  private _onImageLoad = () => {
    this.setState({ isLoading: false })
  }

  renderLoader() {
    return (
      <div className='lightbox__loader'>
        <h2>Loading</h2>
      </div>
    )
  }

  render() {
    const { imageSrc, title, description, onClose } = this.props
    const { isLoading, transition } = this.state

    return (
      <div
        className={cx('lightbox-wrapper', {
          'lightbox-wrapper--enter': transition === 'enter',
          'lightbox-wrapper--leave': transition === 'leave',
        })}
      >
        <div className="lightbox" ref={this._onLightboxRef}>
          <div className="lightbox__content">
            {isLoading && this.renderLoader()}
            <img
              className="lightbox__image"
              src={imageSrc}
              onLoad={this._onImageLoad}
              hidden={isLoading}
            />
            {(title || description) && (
              <div className="lightbox__footer">
                {title && <h3 className="lightbox__title">{title}</h3>}
                {description && (
                  <p className="lightbox__description">{description}</p>
                )}
              </div>
            )}
          </div>
          <button className="lightbox__close" onClick={onClose}>
            <span aria-hidden>&times;</span>
          </button>
        </div>
      </div>
    )
  }
}

export default Lightbox
