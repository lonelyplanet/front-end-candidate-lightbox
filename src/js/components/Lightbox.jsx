import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styleVariables from '@/styles/exports.scss';
import '@/styles/lightbox.scss';

import {
  KEYCODE_ESC,
  KEYCODE_LEFT,
  KEYCODE_RIGHT,
  KEYCODE_DOWN,
  KEYCODE_UP,
  KEYCODE_SPACE,
} from '@/js/constants';


const imageFadeTime = parseInt(styleVariables.imageFadeTimeMs, 10);
const lightboxFadeTime = parseInt(styleVariables.lightboxFadeTimeMs, 10);


class Lightbox extends Component {
  constructor() {
    super();
    this.state = {
      isClosing: false,
      currentImageIndex: 0,
      showImage: false,
    };
    this.handleImageForward = this.handleImageForward.bind(this);
    this.handleImageBack = this.handleImageBack.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.handleKeyUp = this.handleKeyUp.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleImageLoaded = this.handleImageLoaded.bind(this);
    this.setImageIndex = this.setImageIndex.bind(this);
  }

  componentDidMount() {
    document.addEventListener('keyup', this.handleKeyUp);
    document.addEventListener('keydown', this.handleKeyDown);
  }

  componentWillUnmount() {
    document.removeEventListener('keyup', this.handleKeyUp);
    document.removeEventListener('keydown', this.handleKeyDown);
  }

  setImageIndex({ index }) {
    this.setState({
      showImage: false,
    });
    setTimeout(
      this.setState.bind(this, {
        currentImageIndex: index,
      }), imageFadeTime,
      /**
       * Ultimately would be nicer to detect end of css transition animation via event listener,
       * perhaps using $refs to examine the <img> element.
       */
    );
  }

  handleImageBack() {
    const { imageSrcs } = this.props;
    const { currentImageIndex: i } = this.state;
    const previousIndex = (i + imageSrcs.length - 1) % imageSrcs.length;
    this.setImageIndex({ index: previousIndex });
  }

  handleImageForward() {
    const { imageSrcs } = this.props;
    const { currentImageIndex: i } = this.state;
    const nextIndex = (i + 1) % imageSrcs.length;
    this.setImageIndex({ index: nextIndex });
  }

  handleClose() {
    /**
     * "isClosing" state is solely to add a class to the lightbox
     * such that it'll fade out, after which we call the callback from the parent
     * which will presumably set the "isOpen" prop to false.
     * Side note: this would likely be handled in a container (wrapper) component
     * for the lightbox, which has not been implemented here.
     */
    const { onClose: closeCallbackFromParent } = this.props;
    this.setState({ isClosing: true });

    setTimeout(() => {
      this.setState({ isClosing: false });
      closeCallbackFromParent();
    }, lightboxFadeTime);
  }

  handleKeyUp(e) {
    /**
     * Close the lightbox if ESC is pressed,
     * and move through images with left/right arrow keys.
     */
    const { isOpen } = this.props;
    if (! isOpen) {
      return;
    }
    switch (e.keyCode) {
      case KEYCODE_ESC:
        this.handleClose(e);
        break;
      case KEYCODE_LEFT:
        this.handleImageBack();
        break;
      case KEYCODE_RIGHT:
        this.handleImageForward();
        break;
      default:
        // do nothing.
    }
  }

  handleKeyDown(e) {
    /**
     * Prevent scrolling with arrow keys & space bar.
     */
    const { isOpen } = this.props;
    if (! isOpen) {
      return;
    }
    switch (e.keyCode) {
      case KEYCODE_DOWN:
      case KEYCODE_UP:
      case KEYCODE_SPACE:
        e.preventDefault();
        break;
      default:
        // do nothing.
    }
  }

  handleImageLoaded() {
    this.setState({
      showImage: true,
    });
  }

  render() {
    const { isOpen, imageSrcs } = this.props;
    const { currentImageIndex, showImage, isClosing } = this.state;

    return (
      <div
        className={`lightbox${isOpen ? ' active' : ''}${isClosing ? ' is-closing' : ''}`}
        onClick={this.handleClose}
        role="presentation"
      >
        <div
          className="content"
          onClick={e => e.stopPropagation()}
          role="presentation"
        >
          <button
            className="close"
            onClick={this.handleClose}
            type="button"
          >
            x
          </button>
          <button
            className="previous"
            onClick={this.handleImageBack}
            type="button"
          >
            &lt;
          </button>
          <img
            alt=""
            className={showImage ? 'active' : ''}
            src={imageSrcs[currentImageIndex]}
            onLoad={this.handleImageLoaded}
          />
          <button
            className="next"
            onClick={this.handleImageForward}
            type="button"
          >
            &gt;
          </button>
        </div>
      </div>
    );
  }
}

/**
 * As an aside, I like Flow for typechecking and I
 * believe that TypeScript is in use at Lonely Planet,
 * but in the interest of time I'll leave this at propTypes.
 */

Lightbox.propTypes = {
  onClose: PropTypes.func,
  isOpen: PropTypes.bool,
  imageSrcs: PropTypes.arrayOf(PropTypes.string),
};

Lightbox.defaultProps = {
  onClose: () => {},
  isOpen: false,
  imageSrcs: [],
};

export default Lightbox;
