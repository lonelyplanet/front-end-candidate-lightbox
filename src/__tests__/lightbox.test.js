import React from 'react'
import Lightbox from '../App/Lightbox'
import renderer from 'react-test-renderer'

test('Lightbox renders correctly', () => {
  const component = renderer.create(<Lightbox />)

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
