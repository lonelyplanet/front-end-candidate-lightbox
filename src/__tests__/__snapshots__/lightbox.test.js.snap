// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lightbox renders correctly 1`] = `
<div
  className="lightbox"
>
  <div
    className="lightbox-content"
  >
    <img
      src=""
    />
    <h3>
      Loading Catâ€¦
    </h3>
    <a
      className="button"
      href={true}
      onClick={[Function]}
    >
      Close
    </a>
  </div>
</div>
`;
